apps/gendsa.c

```diff
@@ -126,18 +133,40 @@ int gendsa_main(int argc, char **argv)
     if (out == NULL)
         goto end2;
 
-    DSA_get0_pqg(dsa, &p, NULL, NULL);
+    if (!EVP_PKEY_set1_DSA(pkey, dsa)) {
+        BIO_printf(bio_err, "Error setting DSA parameters in PKEY\n");
+        goto end;
+    }
+    DSA_free(dsa);
+    dsa = NULL;
 
-    if (BN_num_bits(p) > OPENSSL_DSA_MAX_MODULUS_BITS)
+    if (EVP_PKEY_bits(pkey) > OPENSSL_DSA_MAX_MODULUS_BITS)
         BIO_printf(bio_err,
                    "Warning: It is not recommended to use more than %d bit for DSA keys.\n"
                    "         Your key size is %d! Larger key size may behave not as expected.\n",
-                   OPENSSL_DSA_MAX_MODULUS_BITS, BN_num_bits(p));
+                   OPENSSL_DSA_MAX_MODULUS_BITS, EVP_PKEY_bits(pkey));
 
     if (verbose)
-        BIO_printf(bio_err, "Generating DSA key, %d bits\n", BN_num_bits(p));
-    if (!DSA_generate_key(dsa))
+        BIO_printf(bio_err, "Generating DSA key, %d bits\n",
+                   EVP_PKEY_bits(pkey));
+
+    if ((ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey, NULL)) == NULL) {
+        BIO_printf(bio_err, "Error initialising PKEY context\n");
         goto end;
+    }
+    EVP_PKEY_free(pkey);
+    pkey = NULL;
+
+    if (!EVP_PKEY_keygen_init(ctx) || !EVP_PKEY_keygen(ctx, &pkey)) {
+        BIO_printf(bio_err, "Error generating key\n");
+        goto end;
+    }
+
+    dsa = EVP_PKEY_get1_DSA(pkey);
+    if (dsa == NULL) {
+        BIO_printf(bio_err, "Error extracting key\n");
+        goto end;
+    }
```

Comment:
Ewwww...  I get wanting to just hack it until it builds, but what you end up doing is creating something that approaches `genpkey` but in a half baked way.  I strongly feel this is the wrong path.

There's been two ideas floating around how to solve this:

1.  When `no-deprecated` (simulating the future removal), simply disable the deprecated sub-command.  Period.  Refer to `genpkey` in this case, possibly as a sub-command shim, but most of all in documentation.
2.  Parse the arguments and rework them for `genpkey`, then call `genpkey_main()` with them.

Reply:
Doesn't this come back to the properly deprecating commands question?

Without this change, the tests don't pass.

Catch 22.

Reply:
Also, this is being deprecated in 3.0 and the changes here can be reverted once the command line deprecation is done.

Reply:
Some of the tests should simply be skipped in this case.  You have the perfect example on how to do so here:

https://github.com/openssl/openssl/blob/8b6ffd40401bd3b78538cb8d496db0c6926185b0/test/recipes/05-test_bf.t#L18-L20

and here:

https://github.com/openssl/openssl/blob/8b6ffd40401bd3b78538cb8d496db0c6926185b0/test/recipes/05-test_cast.t#L19-L21

This isn't fitting for all of the test recipes, there are some where you need to add a `SKIP:` section around the offending commands.

I can do those if you want and push a commit...

Reply:
I'm about to stop coding for the evening.
I can do it in the morning.

Reply:
This is a bit more complicated, the first tests that fail are some of the old SSL tests and only some of the ones being iterated over (inside a single test) do so.  I don't think we've got a good way of expressing this.

Later in the old SSL tests, other things fail because files from the earlier test haven't been properly created.

It's a big mess, but I'll see what I can do.